File: src/pysmo/aimbat/xcorr.f90
!!!---------------------------------------------------------------------!!!
! Filename: xcorr.f90
!   Author: Xiaoting Lou
!    Email: xlou@u.northwestern.edu
!
! Copyright (c) 2009-2012 Xiaoting Lou
!!!---------------------------------------------------------------------!!!
! Module to calculate cross correlation function of two time series with the same length.
!      c(k) = sum_i  x(i) * y(i+k)
!           where k is the time shift of y relative to x.
!Delay time, correlation coefficient, and polarity at maximum correlation are returned.
!
! Added suport to correlation polarity to allow negative correlation maximum.
! Output ccpol=1 if positive or ccpol=-1 if negative. xlou 03/2011
!
! Use numpy.f2py to wrap the code into to a library that can be used in python.
! $ f2py -c -m xcorrf90 xcorr.f90 --fcompiler=gfortran
!
!
!:copyright:
!	Xiaoting Lou
!
!:license:
!	GNU General Public License, Version 3 (GPLv3) 
!	http://www.gnu.org/licenses/gpl.html
!!!---------------------------------------------------------------------!!!


!!!---------------------------------------------------------------------!!!
function crosscorr(x, y, n, k)
! Cross-correlation of array x and y for a given time shift k.
implicit none
integer :: n, k
real(8), dimension(0:n-1) :: x, y
real(8) crosscorr
if (k.ge.0) then
    crosscorr = dot_product(x(0:n-1-k), y(k:n-1))
else
    crosscorr = dot_product(x(-k:n-1), y(0:n-1+k))
endif
end function crosscorr
!!!---------------------------------------------------------------------!!!


!!!---------------------------------------------------------------------!!!
subroutine xcorr_full(x, y, n, shift, delay, ccmax, ccpol)
! Cross-correlation of array x and y.
! Return time shift, correlation coefficient, and polarity at maximum correlation.
implicit none
real(8) :: crosscorr
integer :: n, shift, delay, ccpol
real(8), dimension(0:n-1) :: x, y
real(8) :: cc, ccmax, ccmin
integer :: k, kmin
!f2py intent(in)  :: x, y, n, shift
!f2py intent(out) :: delay, ccmax, ccpol
!f2py intent(hide):: c, k, kmin, ccmin

shift = 1
ccmax = 0
ccmin = 0
kmin = 0
do k = -n+1,n-1,shift
    cc = crosscorr(x,y,n,k)   
    if (cc.gt.ccmax) then
        ccmax = cc
        delay = k
    endif
    if (cc.lt.ccmin) then
        ccmin = cc
        kmin = k
    endif
enddo
if (ccmax.gt.-ccmin) then
    ccpol = 1
else
    ccmax = -ccmin
    delay = kmin
    ccpol = -1
endif
ccmax = ccmax/sqrt( dot_product(x,x) * dot_product(y,y) )

end subroutine xcorr_full
!!!---------------------------------------------------------------------!!!


!!!---------------------------------------------------------------------!!!
subroutine xcorr_fast(x, y, n, shift, delay, ccmax, ccpol)
! Fast cross-correlation using 1 level of coarse shift.
implicit none
integer :: n, shift, delay, ccpol
real(8), dimension(0:n-1) :: x, y
real(8) :: cc, ccmax, ccmin
integer :: s, k, k0, k1, kmin
integer, dimension(0:1) :: shifts
real(8) :: crosscorr
!f2py intent(in)  :: x, y, n, shift
!f2py intent(out) :: delay, ccmax, ccpol
!f2py intent(hide):: cc, s, k, k0, k1, kmin, ccmin

shifts=(/shift,1/)
ccmax = 0
ccmin = 0
kmin = 0
do s = 0,size(shifts)-1
    if (s.eq.0) then
        k0 = 1-n
        k1 = n-1
    else
        k0 = delay-shifts(s-1)
        k1 = delay+shifts(s-1)
    endif
    do k = k0, k1, shifts(s)
        cc = crosscorr(x,y,n,k)
        if (cc.gt.ccmax) then
            ccmax = cc
            delay = k
        endif
        if (cc.lt.ccmin) then
            ccmin = cc
            kmin = k
        endif
    enddo
if (ccmax.gt.-ccmin) then
    ccpol = 1
else
    ccmax = -ccmin
    delay = kmin
    ccpol = -1
endif
enddo

ccmax = ccmax/sqrt( dot_product(x,x) * dot_product(y,y) )

end subroutine xcorr_fast
!!!---------------------------------------------------------------------!!!


!!!---------------------------------------------------------------------!!!
subroutine xcorr_faster(x, y, n, shift, delay, ccmax, ccpol)
! Faster cross-correlation only for time lags around zero.
implicit none
integer :: n, shift, delay, k, kmin, ccpol
real(8), dimension(0:n-1) :: x, y
real(8) :: cc, ccmax, ccmin
real(8) :: crosscorr
!f2py intent(in)  :: x, y, n, shift
!f2py intent(out) :: delay, ccmax, ccpol
!f2py intent(hide):: cc, k, kmin, ccmin

ccmax = 0
ccmin = 0
kmin = 0
do k = -shift, shift, 1
    cc = crosscorr(x,y,n,k)   
    if (cc.gt.ccmax) then
        ccmax = cc
        delay = k
    endif
    if (cc.lt.ccmin) then
        ccmin = cc
        kmin = k
    endif
enddo
if (ccmax.gt.-ccmin) then
    ccpol = 1
else
    ccmax = -ccmin
    delay = kmin
    ccpol = -1
endif
ccmax = ccmax/sqrt( dot_product(x,x) * dot_product(y,y) )

end subroutine xcorr_faster
!!!---------------------------------------------------------------------!!!

File: src/pysmo/aimbat/xcorr.py
#!/usr/bin/env python
#------------------------------------------------
# Filename: xcorr.py
#   Author: Xiaoting Lou
#    Email: xlou@u.northwestern.edu
#
# Copyright (c) 2009 Xiaoting Lou
#------------------------------------------------
"""
Python module to calculate cross-correlation function of two time series with the same length.
      c(k) = sum_i  x(i) * y(i+k)
           where k is the time shift of y relative to x.
Delay time, correlation coefficient, and polarity at maximum correlation are returned.

Added suport to correlation polarity to allow negative correlation maximum.
Output ccpol=1 if positive or ccpol=-1 if negative. xlou 03/2011


:copyright:
    Xiaoting Lou

:license:
    GNU General Public License, Version 3 (GPLv3) 
    http://www.gnu.org/licenses/gpl.html
"""

from numpy import correlate, dot, argmax, argmin, sqrt

def _xcorr(x, y, cmode='full'):
    """ 
    Cross-correlation of two 1-D arrays using 'full' or 'same' mode.
        c[k] = sum_i x[i]*y[i+k]
    Full mode: indices of array c --> j=0:nx+ny-2 <-- k=ny-1:-nx+1:-1 (j=ny-1-k)
    Return time shift of y relative to x at maximum correlation and polarity.
    """
    cc = correlate(x, y, cmode)
    imax = argmax(cc)
    imin = argmin(cc)
    ccmax = cc[imax]
    ccmin = cc[imin]
    if ccmax >= -ccmin:
        ccpol = 1
        delay = len(y)-1-imax
        ccmax =  ccmax / sqrt(dot(x,x)*dot(y,y))
    else:
        ccpol = -1
        delay = len(y)-1-imin
        ccmax = -ccmin / sqrt(dot(x,x)*dot(y,y))
    if cmode == 'same':
        delay -= (len(y)-1)/2
    return delay, ccmax, ccpol

def _xcorr_polarity(x, y, cmode='full'):
    """ 
    Cross-correlation of two 1-D arrays using 'full' or 'same' mode.
        c[k] = sum_i x[i]*y[i+k]
    Full mode: indices of array c --> j=0:nx+ny-2 <-- k=ny-1:-nx+1:-1 (j=ny-1-k)
        It does not correct the polarity, meaning it does only calculate the maximum
        value of cross correlaton, not including minimum, which is different from
        _xcorr
    Return time shift of y relative to x at maximum correlation and polarity.
    """
    cc = correlate(x, y, cmode)
    imax = argmax(cc)
    ccmax = cc[imax]
    ccpol = 1
    delay = len(y)-1-imax
    ccmax =  ccmax / sqrt(dot(x,x)*dot(y,y))
    if cmode == 'same':
        delay -= (len(y)-1)/2
    return delay, ccmax, ccpol

def xcorr_full(x, y, shift=1):
    """ 
    Cross-correlation of two 1-D arrays using 'full' mode.
    Argument shift=1 is here only in order to make the same number of arguments for all xcorr functions.
    """
    return _xcorr(x, y, 'full')

def xcorr_full_polarity(x, y, shift=1):
    """ 
    Cross-correlation of two 1-D arrays using 'full' mode.
    Argument shift=1 is here only in order to make the same number of arguments for all xcorr functions.
    Not correct the polarity
    """
    return _xcorr_polarity(x, y, 'full')

def xcorr_same(x, y):
    """ 
    Cross-correlation of two 1-D arrays using 'same' mode.
    """
    return _xcorr(x, y, 'same')

def xcorr_select(x, y, lags):
    """ 
    Cross-correlation of two time series of the same length 
    for selected lag(shift) times
    """
    n = len(x)
    cc = []
    for k in lags:
        if k >= 0:
            cc.append(dot(x[0:n-k], y[k:n]))
        else:
            cc.append(dot(x[-k:n], y[0:n+k]))
    imax = argmax(cc)
    imin = argmin(cc)
    ccmax = cc[imax]
    ccmin = cc[imin]
    if ccmax >= -ccmin:
        ccpol = 1
        delay = lags[imax]
        ccmax =  ccmax / sqrt(dot(x,x)*dot(y,y))
    else:
        ccpol = -1
        delay = lags[imin]
        ccmax = -ccmin / sqrt(dot(x,x)*dot(y,y))
    return delay, ccmax, ccpol

def xcorr_fast(x, y, shift=10):
    """ 
    Fast cross-correlation of two time series of the same length.
    One level of coarse shift by downsampling the signal.
    """
    sx = x[::shift]
    sy = y[::shift]
    delay, ccmax, ccpol = _xcorr(sx, sy, 'same')
    lags = list(range(delay-shift, delay+shift+1, 1))
    delay, ccmax, ccpol = xcorr_select(x, y, lags)
    return delay, ccmax, ccpol

def xcorr_faster(x, y, shift=10):
    """ 
    Faster cross-correlation only for time lags around zero.
    """
    lags = list(range(-shift, shift+1, 1))
    delay, ccmax, ccpol = xcorr_select(x, y, lags)
    return delay, ccmax, ccpol

def xcorr_select_polarity(x, y, lags):
    """ 
    Cross-correlation of two time series of the same length 
    for selected lag(shift) times
    Do not correct polarity
    """
    n = len(x)
    cc = []
    for k in lags:
        if k >= 0:
            cc.append(dot(x[0:n-k], y[k:n]))
        else:
            cc.append(dot(x[-k:n], y[0:n+k]))
    imax = argmax(cc)
    ccmax = cc[imax]
    ccpol = 1
    delay = lags[imax]
    ccmax =  ccmax / sqrt(dot(x,x)*dot(y,y))
    return delay, ccmax, ccpol

def xcorr_fast_polarity(x, y, shift=10):
    """ 
    Fast cross-correlation of two time series of the same length.
    One level of coarse shift by downsampling the signal.
    Do not correct polarity
    """
    sx = x[::shift]
    sy = y[::shift]
    delay, ccmax, ccpol = _xcorr(sx, sy, 'same')
    lags = list(range(delay-shift, delay+shift+1, 1))
    delay, ccmax, ccpol = xcorr_select_polarity(x, y, lags)
    return delay, ccmax, ccpol

def xcorr_faster_polarity(x, y, shift=10):
    """ 
    Faster cross-correlation only for time lags around zero.
    Do not correct polarity
    """
    lags = list(range(-shift, shift+1, 1))
    delay, ccmax, ccpol = xcorr_select_polarity(x, y, lags)
    return delay, ccmax, ccpol